File: NEETCODE/Easy/validate-parentheses.js
// https://neetcode.io/problems/validate-parentheses

class Solution {
    /**
     * @param {string} s
     * @return {boolean}
     */
    isValid(s) {
        const stack = [];
        const closeSet = { ")": "(", "]": "[", "}": "{" }
        
        for (let i = 0; i < s.length; i++) {
            const char = s[i];
            if (char in closeSet) {
                if (stack.length === 0 || stack.pop() !== closeSet[char]) {
                    return false;
                }
            } else {
                stack.push(char);
            }
        }
        return stack.length === 0;
    }
}

File: NEETCODE/Easy/validate-parentheses.py
# https://neetcode.io/problems/validate-parentheses

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        closeSet = {
            ")": "(", 
            "]": "[", 
            "}": "{"
        }

        for i in s:
            if i in closeSet:
                if stack and stack[-1] == closeSet[i]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(i)
        return True if len(stack) == 0 else False